homeassistant/components/homekit/type_fans.py

```diff
@@ -0,0 +1,159 @@
+                      CHAR_ROTATION_DIRECTION: False,
+                      CHAR_SWING_MODE: False}
+        self._state = 0
+
+        self.chars = []
+        self._features = self.hass.states.get(self.entity_id) \
+            .attributes.get(ATTR_SUPPORTED_FEATURES)
+        if self._features & SUPPORT_SET_SPEED:
+            self.chars.append(CHAR_ROTATION_SPEED)
+            self._speed = None
+        if self._features & SUPPORT_DIRECTION:
+            self.chars.append(CHAR_ROTATION_DIRECTION)
+            self._direction = None
+        if self._features & SUPPORT_OSCILLATE:
+            self.chars.append(CHAR_SWING_MODE)
+            self._oscillating = None
+
+        serv_fan = self.add_preload_service(SERV_FANV2, self.chars)
+        self.char_active = serv_fan.configure_char(
+            CHAR_ACTIVE, value=self._state, setter_callback=self.set_state)
+
+        if CHAR_ROTATION_SPEED in self.chars:
+            speed_list = self.hass.states.get(self.entity_id) \
+                .attributes.get(ATTR_SPEED_LIST)
+            self.min_step = 100 / len(speed_list) if speed_list else 1
+            self.char_speed = serv_fan.configure_char(
+                CHAR_ROTATION_SPEED, value=0,
+                properties={'minStep': self.min_step},
+                setter_callback=self.set_speed)
+
+        if CHAR_ROTATION_DIRECTION in self.chars:
+            self.char_direction = serv_fan.configure_char(
+                CHAR_ROTATION_DIRECTION, value=0,
+                setter_callback=self.set_direction)
+
+        if CHAR_SWING_MODE in self.chars:
+            self.char_swing = serv_fan.configure_char(
+                CHAR_SWING_MODE, value=0, setter_callback=self.set_oscillating)
+
+    def set_state(self, value):
+        """Set state if call came from HomeKit."""
+        _LOGGER.debug('%s: Set state to %d', self.entity_id, value)
+        self._flag[CHAR_ACTIVE] = True
+
+        service = SERVICE_TURN_ON if value == 1 else SERVICE_TURN_OFF
+        self.hass.services.call(self._domain, service,
+                                {ATTR_ENTITY_ID: self.entity_id})
+
+    @debounce
+    def set_speed(self, value):
+        """Set speed if call came from HomeKit."""
+        _LOGGER.debug('%s: Set speed to %d', self.entity_id, value)
+        self._flag[CHAR_ROTATION_SPEED] = True
+        if value != 0:
+            speed_list = self.hass.states.get(self.entity_id) \
+                .attributes.get(ATTR_SPEED_LIST)
+            speed = speed_list[int(round(value / self.min_step)) - 1]
+            self.hass.components.fan.turn_on(
+                self.entity_id, speed=speed)
+        else:
+            self.hass.components.fan.turn_off(self.entity_id)
+
+    def set_direction(self, value):
+        """Set state if call came from HomeKit."""
+        _LOGGER.debug('%s: Set direction to %d', self.entity_id, value)
+        self._flag[CHAR_ROTATION_DIRECTION] = True
+        direction = HOMEKIT_TO_HASS[value]
```

Comment:
Can we use `if` instead? Dictionary mapping is a bit much for just to values.

Reply:
No problem, updated.